script: improve scripted-diff check

max-p-log-p opened the pull request:
Print generic warning if script fails and add extensible way to check for non portable commands. Tested on OpenBSD and Gentoo Linux using the command line with OpenBSD version 6.9 and GNU sed version 4.8 (Gentoo). Checked for POSIX compliance on https://www.shellcheck.net/. Fixes #19815

MarcoFalke commented on test/lint/commit-script-check.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/env sh
```
./ci/lint/06_script.sh: test/lint/commit-script-check.sh: /bin/env: bad interpreter: No such file or directory

4D617278 replied:
I made changes according to the developer notes here: https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#shebang

4D617278 replied:
It should be /usr/bin/env instead of /bin/env

4D617278 replied:
I made changes according to developer notes here: https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#shebang

ajtowns replied:
Missing the `/usr` -- not everyone has merged `/usr`. If  you're changing to to env, probably should also change it to bash as well to avoid getting weird incompatibilities?

4D617278 replied:
I changed it back to #!/bin/sh in a later commit

MarcoFalke commented on test/lint/commit-script-check.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/usr/bin/env sh
```
Missing expected shebang "#!/usr/bin/env bash" or "#!/bin/sh" in test/lint/commit-script-check.sh

MarcoFalke commented on test/lint/commit-script-check.sh:
```diff
@@ -24,10 +24,16 @@ for commit in $(git rev-list --reverse $1); do
         git checkout --quiet $commit^ || exit
         SCRIPT="$(git rev-list --format=%b -n1 $commit | sed '/^-BEGIN VERIFY SCRIPT-$/,/^-END VERIFY SCRIPT-$/{//!b};d')"
+        echo "Script: $SCRIPT"
+        SEDREGEX="^([^#'\"\v]|('[^'\v]*'|\"[^\"\v]*\"))*?(?<![a-z\d])sed(?1)*?-i(?:\.\w+)?[\h,]*(?2)[^#'\"\v]*(?2)"
```
This seems a bit hard to read and impossible to maintain without comments if this should ever fail. I'd say that a simple comment like "If you are using sed, make sure to not use bsd sed syntax" on failure is sufficient.

4D617278 replied:
So you don't need me to check for sed bsd syntax or you want the regex to be simpler? I can ignore comments and strings and revert to an older commit.

4D617278 replied:
Or should I add comments to explain the regex better?

MarcoFalke replied:
I am not sure myself, but I think I'd rather not do the check. It just seems too fragile, especially with the non-greedy regex, which often isn't implemented correctly in regex engines. And how would we check that the `grep --perl-regexp` is using the right regex engine, and not maybe a BSD regex engine? Add another check ...?

4D617278 replied:
Ok, how about I just grep for sed without any regular expressions then? Do you want me to undo my commits or to just add a new one?

MarcoFalke commented on test/lint/commit-script-check.sh:
```diff
@@ -24,6 +24,10 @@ for commit in $(git rev-list --reverse $1); do
         SCRIPT="$(git rev-list --format=%b -n1 $commit | sed '/^-BEGIN VERIFY SCRIPT-$/,/^-END VERIFY SCRIPT-$/{//!b};d')"
+        if echo "$SCRIPT" | grep --quiet "sed"; then
+            echo "Warning: do not use BSD sed syntax"
+            echo "Example: sed -i '' 's/foo/bar/' filename (Bad)"
```
Would be better to only print this warning if the check fails, no?

4D617278 replied:
Which check? The script evaluation?

MarcoFalke replied:
With "check" I mean "script evaluation"

Commit:
Print warning if the script fails and print warning for each line of a nonportable command which is present in the script when the script fails

Fix previous sed command that sometimes fails when using BSD sed, error "expected EOF..."

ajtowns commented on test/lint/commit-script-check.sh:
```diff
@@ -31,7 +31,10 @@ for commit in $(git rev-list --reverse $1); do
-            (eval "$SCRIPT")
+            if ! (eval "$SCRIPT") && (echo "$SCRIPT" | grep --quiet "sed"); then
+                echo "Warning: do not use BSD sed syntax"
+                echo "Example: sed -i '' 's/foo/bar/' filename (Bad)"
```
I would have expected a generic warning that the command failed, maybe something more like:

```sh
if ! (eval "$SCRIPT"); then
    echo "Warning: Script failed (non-zero exit code)"
    if echo "$SCRIPT" | grep -q "sed"; then
        echo "Suggestion: did you use BSD sed syntax?"
        echo "Example: sed -i '' 's/foo/bar/' filename (Bad)"
    fi
fi
```

MarcoFalke replied:
sed doesn't have an exit code for "error", so checking the exit code wouldn't work either way.


```
$ sed -i -e 's/FOOBAR/foobar/g' ./src/init.cpp && echo $?
0

4D617278 replied:
"sed -i '' -e 's/FOOBAR/foobar/g' ./src/init.cpp && echo $?" results in no output for me since the sed command fails and the "echo $?" command is not executed but "sed -i -e 's/FOOBAR/foobar/g' ./src/init.cpp" does in fact work. I think this is because "sed -i -e 's/FOOBAR/foobar/g' ./src/init.cpp" is a legitimate sed command, it just doesn't cause any changes since FOOBAR is not present in the file.

4D617278 replied:
In the OpenBSD 6.9 man pages for sed, it states EXIT STATUS The sed utility exits 0 on success, and >0 if an error occurs.

MarcoFalke replied:
failing to replace is not an error

4D617278 replied:
The line "git --no-pager diff --exit-code $commit && echo "OK" || (echo "Failed"; false) || RET=1" should check if the script failed to change anything, do you want to check each command and error if if has no change?

4D617278 replied:
Do you want to not check for non zero exit codes but only check if the output of the script is the same as the commit?

78051301012 reviewed (commented):
.

adamjonas commented:
Hi @4D617278, would you mind squashing your commits as described in [these guidelines](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits)?

MarcoFalke commented:
Looks like #https://github.com/bitcoin/bitcoin/issues/19815 is already fixed?
